File: backend/src/main/java/com/voltpay/backend/service/ContactsBackupService.java
package com.voltpay.backend.service;

import com.voltpay.backend.model.ContactBackup;
import com.voltpay.backend.repository.ContactsBackupRepository;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.List;
import java.util.Optional;

@Service
public class ContactsBackupService {

    private final ContactsBackupRepository repository;

    public ContactsBackupService(ContactsBackupRepository repository) {
        this.repository = repository;
    }

    public ContactBackup syncContacts(String phoneNumber, List<ContactBackup.ContactEntry> contacts) {
        Optional<ContactBackup> existingOpt = repository.findByPhoneNumber(phoneNumber);
        ContactBackup backup = existingOpt.orElse(new ContactBackup());
        
        backup.setPhoneNumber(phoneNumber);
        // Simple strategy: overwrite with client's list. 
        // In a real app we might do true merging, but client sync logic merges before sending.
        backup.setContacts(contacts);
        backup.setLastSyncedAt(Instant.now());
        
        return repository.save(backup);
    }

    public Optional<ContactBackup> getContacts(String phoneNumber) {
        return repository.findByPhoneNumber(phoneNumber);
    }
}

File: backend/src/main/kotlin/com/voltpay/backend/service/ContactsBackupService.kt
package com.voltpay.backend.service

import com.voltpay.backend.model.ContactBackup
import com.voltpay.backend.repository.ContactsBackupRepository
import org.springframework.stereotype.Service
import java.time.Instant

@Service
class ContactsBackupService(private val repository: ContactsBackupRepository) {

    fun syncContacts(phoneNumber: String, contacts: List<ContactBackup.ContactEntry>): ContactBackup {
        val backup = repository.findByPhoneNumber(phoneNumber).orElse(ContactBackup())
        backup.phoneNumber = phoneNumber
        backup.contacts = contacts
        backup.lastSyncedAt = Instant.now()
        return repository.save(backup)
    }

    fun getBackup(phoneNumber: String): ContactBackup? {
        return repository.findByPhoneNumber(phoneNumber).orElse(null)
    }
}
